src/transforms.rs

```diff
@@ -167,15 +169,70 @@ pub enum MessageTransform {
     ExpressionTransform(Expression<'static>),
 }
 
-pub struct Transformer {
-    transforms: HashMap<String, MessageTransform>,
+pub struct ValuePath {
+    parts: Vec<String>,
 }
 
-impl Transformer {
-    pub fn new(transforms: HashMap<String, MessageTransform>) -> Self {
-        Self { transforms }
+impl ValuePath {
+    fn from_str(path: &str) -> Self {
+        let parts: Vec<String> = path.split(".").map(|s| s.to_string()).collect();
+
+        ValuePath { parts }
+    }
+
+    fn part_at(&self, index: usize) -> Option<&str> {
+        self.parts.get(index).map(|s| s.as_ref())
+    }
+
+    fn len(&self) -> usize {
+        self.parts.len()
+    }
+}
+
+fn set_value(object: &mut Map<String, Value>, path: &ValuePath, path_index: usize, value: Value) {
+    match value {
+        // Don't set if the extracted value is null.
+        Value::Null => { /* noop */ }
+        _ => {
+            let part = path.part_at(path_index);
+
+            if let Some(property) = part {
+                if path_index == path.len() - 1 {
+                    // this is the leaf property - set value on the current object in context.
+                    object.insert(property.to_string(), value);
+                } else {
+                    if let Some(next_o) = object
+                        .get_mut(property)
+                        .map(|v| v.as_object_mut())
+                        .flatten()
+                    {
+                        // the next object already exists on the object. recurse.
+                        set_value(next_o, path, path_index + 1, value);
+                    } else {
+                        // this is not the leaf property and the parent object does not exist yet.
+                        // create an object, then recurse.
+                        let next_o = Value::Object(Map::new());
+                        object.insert(property.to_string(), next_o);
+                        set_value(
+                            &mut object.get_mut(property).unwrap().as_object_mut().unwrap(),
+                            path,
+                            path_index + 1,
+                            value,
+                        );
```

Comment:
I don't think the unwrap will ever fail here, but still worth rewriting the order to get rid of it:

```suggestion                    
                        set_value(
                            &mut next_o,
                            path,
                            path_index + 1,
                            value,
                        );
                        object.insert(property.to_string(), next_o);
```